core/src/SYCL/Kokkos_SYCL_Instance.cpp

```diff
@@ -130,14 +141,21 @@ void SYCLInternal::initialize(const sycl::device& d) {
 
 void SYCLInternal::finalize() {
   SYCL().fence();
-  was_finalized = 1;
+  was_finalized = true;
   if (nullptr != m_scratchSpace || nullptr != m_scratchFlags) {
     // FIXME_SYCL
     std::abort();
   }
 
   m_indirectKernelMem.reset();
   m_indirectReducerMem.reset();
+  // guard erasing from all_queues
+  {
+    std::lock_guard<std::mutex> lock(mutex);
+    all_queues.erase(
+        std::remove(all_queues.begin(), all_queues.end(), &m_queue),
```

Comment:
Why do you remove all instead of the first occurrence?  That is a bug if someone creates two execution spaces from the same queue.

Reply:
Currently, we are copying the `sycl::queue` given to a `std::optional<sycl::queue>` so that the pointers we store should be distinct. Anyway, it is sufficient to only search for the first match.